IFP: Multi height cell support

mina1460 opened the pull request:
This PR would serve as the foundation for supporting multi-height cells.

It also includes support for removing the need for the -site name flag for init_floorplan command. It would still work with and without it.

maliberty commented on src/ifp/include/ifp/InitFloorplan.hh:
```diff
@@ -35,8 +35,9 @@
+#include <set>
 #include <string>
-
+#include <vector>
```
I don't see a usage of vector in this header.  It is needed elsewhere please add it to the file(s) that use it.

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
+    // get the site of the master
+    auto site = master->getSite();
+    if (site == nullptr) {
+      continue;
```
A warning would be good here as this would be unexpected.

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
+  });
+
+  // smallest site height
+  double min_site_height = site_vec[0]->getHeight();
+  double factor = 1.0;
+  for (auto site : site_vec) {
+    double site_height = site->getHeight();
+    factor = site_height / min_site_height;
+    // check if the site height is a multiple of the smallest site height
+    if (std::ceil(factor) != std::floor(factor)) {
```
Use integers here not doubles.  Check with the modulo operator not ceil/floor.

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
+    if (std::ceil(factor) != std::floor(factor)) {
+      logger_->error(
+          IFP,
+          228,  // TODO: add error code
```
Why is this a TODO?

mina1460 replied:
I didn't know how I should pick an error code, I thought I should document the code somewhere as each code seems to be unique for a message. Should I just pick the first available number?

maliberty replied:
Use the next available from ifp/messages.txt (ie 40)

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
+      logger_->error(
+          IFP,
+          228,  // TODO: add error code
+          "site height {} is not a multiple of the smallest site height {}",
```
Give the site names too

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
+          site_height,
+          min_site_height);
+      // erase from the set if the site height is not a multiple of the smallest
+      sites.erase(site);
```
you'll never get here as error() throws an exception.  The erase wouldn't be safe while iterating anyways

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
+  return sites;
+}
+
 void InitFloorplan::makeRows(dbSite* site,
```
Please add a unit test that inserts double height rows.

maliberty replied:
This still needs addressing

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -360,11 +389,53 @@ void InitFloorplan::updateVoltageDomain(dbSite* site,
+  });
+
+  // smallest site height
+  double min_site_height = site_vec[0]->getHeight();
```
site_vec[0] is invalid if sites.empty()

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -166,11 +167,41 @@ void InitFloorplan::initFloorplan(const odb::Rect& die,
+  // and user didn't specify a site that we could find
+  if (sites.empty()) {
+    logger_->error(IFP, 33, "No sites found.");
+    return;
```
rm, not reached

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -166,11 +167,41 @@ void InitFloorplan::initFloorplan(const odb::Rect& die,
+    return;
+  }
+  // get min site height using std::min_element
+  int min_site_height
```
const

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -166,11 +167,41 @@ void InitFloorplan::initFloorplan(const odb::Rect& die,
+            ->getHeight();
+
+  eval_upf(logger_, block_);
+  for (auto site : sites) {
+    if (core.xMin() >= 0 && core.yMin() >= 0) {
```
The core size won't change with sites so this check should be hoisted.

maliberty commented on src/ifp/src/InitFloorplan.cc:
```diff
@@ -375,8 +446,9 @@ void InitFloorplan::makeRows(dbSite* site,
   for (int row = 0; row < rows_y; row++) {
-    dbOrientType orient = (row % 2 == 0) ? dbOrientType::R0   // N
-                                         : dbOrientType::MX;  // FS
+    dbOrientType orient = (factor == 2 or row % 2 == 0)
```
any even factor not just 2 (eg quad height cells)

QuantamHD commented:
@mina1460 Would you mind rebasing to master I introduced a clang-tidy action that leaves comments on your PR that I would like to see here https://github.com/The-OpenROAD-Project/OpenROAD/pull/2980

mina1460 commented:
@QuantamHD Done

QuantamHD commented:
Thanks!

maliberty commented on src/ifp/test/mock.tcl:
```diff
@@ -0,0 +1,7 @@
+read_def mock.def
+initialize_floorplan -utilization 30 \
+  -aspect_ratio 0.5 \
+  -core_space 1 
```
The test needs to do something to validate the results.  Likely you want a write_def and a .defok file.  Please see other tests for examples.

mina1460 replied:
yes, absolutely. I just wanted to verify that the lef/def files and the mock macros are OK and this is indeed testing a double height cell before spending time on a defok file.

Should I proceed with it?

maliberty replied:
yes

maliberty commented on src/ifp/test/mock.def:
```diff
@@ -5,7 +5,7 @@ DESIGN simple_design ;
 DIEAREA ( 0 0 ) ( 2000 2000 ) ;
 ROW ROW_0 SingleHeightSite 0 0 FS DO 10 BY 1 STEP 1000 0 ;
-ROW ROW_1 DoubleHeightSite 1200 2200 N DO 12 BY 1 STEP 1100 0 ;
+ROW ROW_1 DoubleHeightSite 0 1000 N DO 10 BY 1 STEP 1000 0 ;
```
Why are there rows in the input DEF?  The goal of ifp is to build the rows.  You could just read a .v for this test which would better mirror ORFS (ie the yosys output).

mina1460 replied:
makes sense. I will fix it

QuantamHD commented:
@maliberty @vvbandeira I think @mina1460 is getting a high number of notifications from the sync upstream, because of the waiting for approval status. We should probably add him to the list of approved workflow runners.

maliberty commented:
@QuantamHD  once his first PR is merged he will automatically be added.  I've never had to do it manually.

maliberty commented:
The CI failures need to be addressed before this can be merged

mina1460 commented:
Hi @maliberty 

All CI/CD failures are now resolved. 

I modified the updateVoltageDomain and removed the site from it so it is no longer needed. As you said, this had the potential of cutting double-height rows into single-height rows, but I don't think this is a problem with this modification anymore.
